File: ps0/hw0.py
import sys
import os
import re
## CS 458/558
## Fall 2015
## hw0.py
## Aleksandra Zakrzewska



## Function 1: palindrome

def palindrome(filename = "/usr/share/dict/words", letter = "all"):
	# open file
	file = open(filename, 'r')

	palindromes = []

	for line in file:
		#skip any words that do not begin with the correct letter--upper or lower case
		i =0
		while (not(line[i].isalpha())):
			#print line[i], "is not alpha"
			i+=1
			
		if ((letter!= "all" and (line[i].lower()!=letter.lower()))):
			#print line[i], letter
			continue
		length = len(line);
		j =length-2
		#check if all letters are reversed at the end
		word = line[:length-1]
		while i<j:
			#print word[i], word[j]
			if ( ((not (word[i].isalpha())) and (not(word[i].isdigit()))) or (word[i] == ' ')):
				i+=1
			elif ( ((not (word[j].isalpha())) and (not(word[j].isdigit()))) or (word[j] == ' ')):
			#elif ( (not (word[j].isalpha()) or word[i].isdigit()) or (word[i] == ' ')):
				j-=1
			elif ((word[i].lower() != word[j].lower()) and (ord(word[i]) != ord(word[j]) -32)):
				#print word[i].lower(), word[j].lower(),  "not same"
				break
			else:
				#print "were same"
				i+=1
				j-=1
		else: 
			print word
			palindromes.append(word)
	#print palindromes



# >>> palindrome(letter = "b")
# B
# b
# Bab
# bab
# B/B
# BB
# bb
# BBB
# Beeb
# Bib
# bib
# Bob
# bob
# boob
# Bub
# bub



## Function 2: anagram

# Example output
# >>> anagram()
# ['alerts', 'alters', 'artels', 'estral', 'laster', 'lastre', 'rastle', 'ratels', 'relast', 'resalt', 'salter', 'slater', 'staler', 'stelar', 'talers']
# 
# >>> anagram(target = 'male')
# ['alem', 'alme', 'amel', 'lame', 'leam', 'male', 'meal', 'mela']

def anagram(filename = "/usr/share/dict/words", target = False):
	
	file = open(filename, 'r')

	anagrams = {}

	maxcount=0
	maxset=""

	for line in file:
		#for every word in the file, count the number of each letter 
		freq= [0] *27 
		word = line[:len(line)-1]
		for i in range (len(word)):
			if ( not (word[i].isalpha())):
				continue
			freq[ord(line[i].lower())-97]+=1
		freqtup= tuple(freq)
		#use that set as a key to store the word in a dict
		if freqtup not in anagrams:
				anagrams[freqtup]=[0, []]
		anagrams[freqtup][0]+=1
		anagrams[freqtup][1].append(word)
		#keep track of which key has the highest count of words
		if anagrams[freqtup][0]>maxcount:
			maxset=freqtup
			maxcount = anagrams[freqtup][0]
	
	#find the target key and print the set of words with that key
	if target!=False:
		targetSet= [0] *27 
		target=target.lower()
		
		for i in range (len(target)):
			if ( not (target[i].isalpha())):
				continue
			targetSet[ord(target[i])-97]+=1
		targetSet= tuple(targetSet)
		for key in anagrams[targetSet][1]:
			print key

	#print the set of words belonging to the most counted key 
	else :
		#print anagrams[maxset][1]
		for key in anagrams[maxset][1]:
			print key


#call the appropriate function with the appropriate argument
def main(arg):
	if len(arg)<1:
		print ("Usage: hw0.py anag|palin [target | letter] [filename = filename]")
		exit()
		#print arg
	if arg[0] == "anagram":
		target = False
		if len(arg)>=2:
			target = arg[1]
		anagram( target = target)

	elif arg[0] == "palindrome":
		letter="all"
		if len(arg)>=2:
			letter = arg[1]
		palindrome( letter = letter)


if __name__ == "__main__":
   main(sys.argv[1:])

File: ps0/hw0.R
## CS 458/558
## Fall 2015
## hw0.R
#install.packages("pryr")
#library("pryr")
library("stringr")
#library("hash")


# write a function palindrome which reads an ascii file comprising one word
# per line and returns all the palindromes.  
# The default file is the standard UNIX online dictionary: /usr/share/dict/words
# There is another optional argument of a single letter, in which case the
# program returns only palindromes which start with that letter.

## Use the following function signature. 
palindrome <- function(filename = "/usr/share/dict/words", letter = "all"){
  file<- readLines(filename)
  #print( file)
  for (i in 1:length(file)) {
    word<-file[i]
    lower<-tolower(word)
    lower = str_replace_all(lower, "[^[:alnum:]]", "")
    lower<-strsplit(lower, "")[[1]]
    #print(word)
    #print(lower)
    if (letter != "all" && lower[1]!= letter) {
      next
    }
    len <- length(lower)
    #print(word)
    i <- 1
    j <- len
    while (i<=j) {
      #print(lower[i])
      #print(lower[j])
      if (lower[i] != lower[j] ) {
        #print("do not match")
        break
      }
      i<-i+1
      j<-j-1
    }
    if (j<=i) {
      print(word)
    }
  }
}
  

## Function 2: anagram
##
# write a function anagram which reads an ascii file comprising one word
# per line and returns all the largest set of words which are anagrams.  
# The default file is the standard UNIX online dictionary: /usr/share/dict/words
# (During development, you might want to use a smaller dictionary.)
# 
# There is another optional argument which is a single word.  
# In that case, the function returns all anagrams for that word 

# Example output
# > anagram()
# [1] "alerts" "alters" "artels" "estral" "laster" "lastre" "rastle" "ratels" "relast" "resalt"
# [11] "salter" "slater" "staler" "stelar" "talers"
# 
# > anagram(target = 'male')
# [1] "alem" "alme" "amel" "lame" "leam" "meal" "mela"

## Use the following function signature. 
anagram <- function(filename = "/usr/share/dict/words", target = FALSE){
 dict <- new.env()
 countdict <- new.env()
 maxcount = 0
 maxset = ""
 
 file<- readLines(filename)
 #print( file)
 for (i in 1:length(file)) {
   word <- file[i]
   freq <- rep(c(0), 26)
   lower<-tolower(word)
   lower = str_replace_all(lower, "[^[:alnum:]]", "")
   lower<-strsplit(lower, "")[[1]]
   len <- length(lower)
   for (i in 1:len) {
     letter <- strtoi(charToRaw(lower[i]),16L) -96
     freq[letter]<-freq[letter]+1
   }
   freq= str_replace_all(paste(toString(freq), collapse=NULL), "[ ,]" ,"")
   #print(word)
   #print(freq)
   if (!(exists(freq, envir=dict))) {
     countdict[[freq]] <- 0
     dict[[freq]] <- ""
     }
   count = get(freq, envir=countdict) + 1
   countdict[[freq]] <- count
   
   set = append(get(freq, envir = dict), word)
   dict[[freq]]<-set
   #print(set)
   
   if (count> maxcount) {
     maxcount <- count
     maxset <-freq
   }
 }
 if (target == FALSE) {
   most=get(maxset, envir = dict)
   for (i in 2:length(most)) {
     print (most[i])
   }
 }
 else {
   freq <- rep(c(0), 26)
   lower<-tolower(target)
   lower = str_replace_all(lower, "[^[:alnum:]]", "")
   lower<-strsplit(lower, "")[[1]]
   len <- length(lower)
   for (i in 1:len) {
     letter <- strtoi(charToRaw(lower[i]),16L) -96
     freq[letter]<-freq[letter]+1
   }
   target= str_replace_all(paste(toString(freq), collapse=NULL), "[ ,]" ,"")
   if (exists(target, envir=dict) ) {
     anagrams = get(target, envir = dict)
     for (i in 2:length(anagrams)){
       print(anagrams[i])
     }
   }
 }
}


## Bonus question (no credit)
## An auto-antonym is a word that has two meanings that are opposites.
## For example, "dust" can mean to remove dust (I am dusting under the bed)
## and also "dust" can mean to add dust (I am dusting the cake with powdered sugar)
##
## How would you use a computer to come up with a list of auto-antonyms?
## Hint: there is more than one answer.